### ISSUE TITLE
Collaborator Gating Based On Label

### ISSUE THREAD

**Issue by 0x4007** [ubiquity-os-marketplace/command-start-stop#62]
- Since we still are ways off to implementing XP, we should implement a simpler gating system. 
- Similar to how we manage max concurrent tasks per role, we should also have a config to gate based on the label. 
- There's no point in making a config for admin.

```yml
  - uses:
    - plugin: https://ubiquibot-assistive-pricing-development.ubiquity.workers.dev
      with:
        labels:
          time:
            - "Time: <1 Hour"
              collaboratorOnly: false
            - "Time: <2 Hours"
              collaboratorOnly: false
            - "Time: <4 Hours"
              collaboratorOnly: true
            - "Time: <1 Day"
              collaboratorOnly: true
            - "Time: <1 Week"
              collaboratorOnly: true
            - "Time: <1 Month"
              collaboratorOnly: true
          priority:
            - "Priority: 1 (Normal)"
              collaboratorOnly: false
            - "Priority: 2 (Medium)"
              collaboratorOnly: false
            - "Priority: 3 (High)"
              collaboratorOnly: true
            - "Priority: 4 (Urgent)"
              collaboratorOnly: true
            - "Priority: 5 (Emergency)"
              collaboratorOnly: true
        basePriceMultiplier: 2
        publicAccessControl:
          setLabel: true
          fundExternalClosedIssue: false
```

We could even embed metadata in the label description, although for this purpose maybe via the config here is simple enough.

**Comment by 0x4007** [comment]
@gentlementlegen I think this could save us from a lot of unnecessary blockers. Perhaps you can prioritize it?

**Comment by gentlementlegen** [comment]
@0x4007 Should we add a tag on the task or should it just be taken from the configuration?

**Comment by 0x4007** [comment]
Just from the configuration. I don't think extra tags is a good idea.

**Comment by gentlementlegen** [comment]
@0x4007 Any message to display like "This issue is for collaborators only?"

**Comment by 0x4007** [comment]
In the future I think its worthwhile to assume these small details that are fast to change so that you are not held up. This message seems fine to start!

**Comment by gentlementlegen** [comment]
@0x4007 I just realized this was created under the configuration of `assistive-pricing`, so I think we should create a new label on the issue for this so `command-start-stop` can check it and know if the task is restricted or not.

**Comment by 0x4007** [comment]
>  I don't think extra labels is a good idea.

**Comment by gentlementlegen** [comment]
First idea then would be to add a description inside the label saying "Collaborator only" so that `command-start-stop` can check the restrictions on the labels, without having to add new ones.

**Comment by 0x4007** [comment]
> We could even embed metadata in the label description, although for this purpose maybe via the config here is simple enough.

I don't think it makes sense to implement a new approach. Just read from config like we always do because it works fine. We can embed metadata in the description if we have more complex needs.

**Comment by gentlementlegen** [comment]
The configuration you linked is not for this plugin, plugins are not aware of the configuration of the others which is what I am trying to solve here.

**Comment by 0x4007** [comment]
Then you have to sync. It's more complicated than just reading from config every time I think

**Comment by gentlementlegen** [comment]
What do you mean by "to sync"?

**Comment by 0x4007** [comment]
Any changes to the config have to sync all the descriptions every time with your proposal. It's more complex.

**Comment by 0x4007** [comment]
Then let's embed in the description if that means less new code.

**Comment by gentlementlegen** [comment]
Will depend on https://github.com/ubiquity-os-marketplace/daemon-pricing/pull/18 to be completed in order to have the label updates working properly.

**Comment by gentlementlegen** [comment]
So through the UI the labels would look like this
![image](https://github.com/user-attachments/assets/5cb3b24d-e7a5-4412-907e-045b32e24656)

which I think is helpful when creating the tasks.

**Comment by 0x4007** [comment]
Perhaps we can use emoji so that it stands out like ⚠️

**Comment by gentlementlegen** [comment]
Not sure if GitHub supports that, will have a look.

**Comment by 0x4007** [comment]
To clarify I mean in the description. ⚠️ contributor only.

**Comment by gentlementlegen** [comment]
Yes got it. It seems to be supported:
![image](https://github.com/user-attachments/assets/cf13e126-6b93-461b-a982-fb8079ddefeb)

**Comment by 0x4007** [comment]
Price shoudn't have that

**Comment by gentlementlegen** [comment]
@0x4007 Yes it was for testing, I didn't remove that tag, pricing won't have them.

**Comment by gentlementlegen** [comment]
Still waiting on https://github.com/ubiquity-os-marketplace/daemon-pricing/pull/18 to be merged in.

**Comment by Keyrxng** [comment]
https://github.com/ubiquity-os-marketplace/daemon-pricing/pull/42#issuecomment-2431077191

Since https://github.com/ubiquity-os-marketplace/daemon-pricing/pull/18 is not actually blocking this and after looking over things I'd like offer a super simple ultra low code suggestion that can be merged in under an hour


> To me it seems like a lazy answer to keep adding labels for every feature. We already went through this with v1 of bot and the issues all looked like a mess.

> Then you have to sync. It's more complicated than just reading from config every time I think

It really is as easy as reading from the config when the user calls `/start`, then if the task labels are lower than the new config thresholds they can start it.

```yml
   - uses:
     - plugin: http://localhost:4000
       runsOn:  ["issue_comment.created", "issues.assigned", "issues.unassigned", "pull_request.opened", "pull_request.edited"]
       with:
         reviewDelayTolerance: "1 Day"
         taskStaleTimeoutDuration: "30 Days"
         maxConcurrentTasks: 
           member: 2
           contributor: 5
         startRequiresWallet: false
         rolesWithReviewAuthority: ["OWNER", "ADMIN", "MEMBER", "CONTRIBUTOR"]
        
         collaboratorOnlyThresholds:{
           time: 1 week,
           priority: 4 urgent
         }
```

The benefits of using `assistive-pricing` I guess is being able to update label descriptions but that was never a concern before, as a comment on the task is usually the chosen approach to user feedback as opposed to bury inside label info which a contributor may not see unless they open up the label selection dropdown.

"time" and "priority" prefixes may be changed by partners I think but not us so can deal with that later if that's an issue reading this approach.

**Comment by Keyrxng** [comment]
example, I can't set labels in this org so I cannot see all of the available labels, meaning this description is hidden to contributors, so if they use `/start` we'll need to provide feedback anyway.

![image](https://github.com/user-attachments/assets/e164e824-0a7b-4d4e-9e8c-d3fbe265a8f8)

**Comment by gentlementlegen** [comment]
@Keyrxng Thanks for the idea! However this is all already implemented [there](https://github.com/ubiquity-os-marketplace/command-start-stop/pull/65) so I could change this after we merge https://github.com/ubiquity-os-marketplace/daemon-pricing/pull/42

**Comment by Keyrxng** [comment]
> @Keyrxng Thanks for the idea! However this is all already implemented [there](https://github.com/ubiquity-os-marketplace/command-start-stop/pull/65) so I could change this after we merge [ubiquity-os-marketplace/daemon-pricing#42](https://github.com/ubiquity-os-marketplace/daemon-pricing/pull/42)

I saw that PR was merged yeah, although if the descriptions are only visible to people with access it's more of an internal quality of life having label descriptions since only contributors can see them.

Since they are not dependant on each other the change could be made in `start-stop` first and not be held back by me or any other bugs or whatever, if the plan is to change it later, may as well do it first as there is nothing holding that back. 

Sorry, just trying to provide value/solutions vs problems and delays considering how things are rn, hope you understand.
